pkg/monitor/monitor.go

```diff
@@ -75,67 +74,36 @@ func (m *Monitor) Run(ctx context.Context) (retErr error) {
 
 	m.Logger.Info("Initialized ServiceDescriptor for synthetic checks")
 
-	// Wait until the corresponding ServiceInstance is created.
-	// Sleeps for a short period of time, which for now should be enough I think
-	time.Sleep(m.ExpectedProcessingTime)
-
-	err = m.verifyUpsServiceInstance(ctx, m.ServiceDescriptorName, m.Version)
+	err = wait.PollImmediate(pollDelay, m.ExpectedProcessingTime, m.verifyServiceDescriptorStatus(sd.ObjectMeta.Name))
 	if err != nil {
-		m.Logger.Error("ServiceInstance verification has failed", zap.Error(err))
 		return err
 	}
 
-	m.Logger.Info("ServiceInstance verification has succeeded")
+	m.Logger.Info("ServiceDescriptor Status verification has succeeded")
 	return nil
 }
 
-func (m *Monitor) verifyUpsServiceInstance(ctx context.Context, namespace, expectedVersion string) error {
-	si, err := m.serviceInstance(ctx, string(resourceName), namespace)
-	if err != nil {
-		return errors.Wrapf(err, "could not get ServiceInstance %q", resourceName)
-	}
-
-	err = verifySpec(si.Spec, namespace, expectedVersion)
-	if err != nil {
-		return errors.WithMessage(err, "unexpected Spec")
-	}
+func (m *Monitor) verifyServiceDescriptorStatus(name string) func() (bool, error) {
+	return func() (bool, error) {
+		var sd *comp_v1.ServiceDescriptor
 
-	err = verifyStatus(si.Status)
-	if err != nil {
-		return errors.WithMessage(err, "unexpected Status")
-	}
-
-	return nil
-}
-
-func verifySpec(spec v1beta1.ServiceInstanceSpec, namespace, expectedVersion string) error {
-	var data map[string]string
-	err := json.Unmarshal(spec.Parameters.Raw, &data)
-	if err != nil {
-		return errors.Wrapf(err, "could not unmarshal spec's parameters")
-	}
+		sd, err := m.ServiceDescriptorClient.Get(name, meta_v1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
 
-	foundVersion, ok := data[versionParameter]
-	if !ok {
-		return errors.Errorf("version parameter missing from ServiceInstance %q in %q", resourceName, namespace)
-	}
-	if foundVersion != expectedVersion {
-		return errors.Errorf("found ServiceInstance with embedded version %q but expecting %q", foundVersion, expectedVersion)
-	}
+		_, ready := cond_v1.FindCondition(sd.Status.Conditions, cond_v1.ConditionReady)
+		if ready == nil {
+			// Probably fine, this can trigger before the SD is processed by the cluster
+			return false, nil
+		}
```

Comment:
generation and observed generation should be checked here in the future.

Reply:
@ash2k SD status still doesn't have `ObservedGeneration` AFAICT :(

Reply:
yes, it does not. That is why I'm saying "in the future" :)